src/compat.rs

```diff
@@ -18,12 +19,25 @@ use strum_macros::{EnumCount as EnumCountMacro, EnumIter};
 /// a moving target that would change the semantics of your Landlock rule
 /// when migrating to a newer version of this crate
 /// (i.e. non-breaking change with new supported features).
-/// This usage should then be considered indeterministic because requested features
+/// This usage is then an undefined behavior because requested features
```

Comment:
I don't see how this would be undefined behavior? The behavior is 100% predictable and consistent based on the `rust-landlock` version being used, right? There's no dragons to be found anywhere.

Updating landlock can change the behavior of the application, but that's just how updating dependencies works in general. I don't think that's unexpected and might even be intentional? I know that's why I personally used `BitFlags::all`.

Reply:
It may be undefined because when app's code is written, rust-landlock might be at v1.2.0 and when this lib will add support for a new access right it will bump to v1.3.0 which will be automatically used by someone building the app. At this point, folks building the app may know or not what the new access right is, but there is a (big) chance that they don't know. The impact will be that the built app, with the same source code, will have a different behavior and may break if a required kernel feature is then denied because of the `BitFlags::ALL`. So, at the time of the code was written this behavior couldn't be known, which I translate to "undefined".

If people use `BitFlags::ALL` their is a good chance that their application will break at some point because of that.

Reply:
> It may be undefined because when app's code is written, rust-landlock might be at v1.2.0 and when this lib will add support for a new access right it will bump to v1.3.0 which will be automatically used by someone building the app.

Lockfiles should prevent this from ever happening.

Reply:
Do you mean `Cargo.lock`? This shouldn't be used for libraries (that might use rust-landlock), and even apps should not rely on that (in theory) to not break.

Reply:
I was chatting about this with @l0kod offline.  Here's my two cents on the undefined behavior discussion.

IMHO it is correct that this is "undefined behavior".  The discussion seems to dance around exactly what is meant by undefined behavior, but in the sense that "the spec doesn't define it", I think this is undefined.  The point you are making (IIUC) is that a future landlock may make unexpected changes.  You are not making any guarantee about future access rights being added.

I do think that it would be a stretch to call such behavior "indeterministic".  The future behavior will presumably be deterministic (non-deterministic security decisions sure sounds like a security bug in most contexts...).

All of that said, I think the reason "undefined behavior" can feel a little weird here is because in practice you do expect a pretty constrained set of behaviors.  Either access will be allowed or denied.  Possibly in ways that are difficult to anticipate at this point.  "We don't define whether we allow or deny an access" is very different in practice from "We don't define anything, and setting your computer on fire is legal behavior".  In some contexts (eg compilers) it can be genuinely difficult to predict the effect of optimizations on undefined behavior.  As [this article](https://pvs-studio.com/en/blog/posts/cpp/0938/) points out, in the presence of optimizations, it's theoretically possible for C compilers to reorder instructions around undefined behavior in ways that cause writes at arbitrary locations in the program's memory space, which very much get outside the scope of things like "crash" that C programmers tend to think that a program with undefined behavior may do.

The reason I bring all that up, is that the landlock scenario is much less scary (I assume) than a compiler that does optimizations such that the behavior could in theory be extremely destructive.  It's just not defined what access will be allowed or disallowed in future versions if you rely on future access controls.  For that reason, while "undefined behavior" is technically correct, I think it might be helpful to use a less loaded phrase such as "landlock makes no guarantees around access decisions...", which explicitly caveat the scope of the lack of definition.  It's both more precise and less scary that way.

Reply:
Thanks @dburgener, that's helpful! I agree and I'll update the doc accordingly.